dispatch/dispatch_test.go

```diff
@@ -568,6 +569,56 @@ route:
 	require.Len(t, alertGroups, 6)
 }
 
+func TestGroupsAlertLCObserver(t *testing.T) {
+	confData := `receivers:
+- name: 'testing'
+
+route:
+  group_by: ['alertname']
+  group_wait: 10ms
+  group_interval: 10ms
+  receiver: 'testing'`
+	conf, err := config.Load(confData)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	logger := log.NewNopLogger()
+	route := NewRoute(conf.Route, nil)
+	marker := types.NewMarker(prometheus.NewRegistry())
+	alerts, err := mem.NewAlerts(context.Background(), marker, time.Hour, nil, logger, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer alerts.Close()
+
+	timeout := func(d time.Duration) time.Duration { return time.Duration(0) }
+	recorder := &recordStage{alerts: make(map[string]map[model.Fingerprint]*types.Alert)}
+	m := NewDispatcherMetrics(true, prometheus.NewRegistry())
+	observer := alertobserver.NewFakeAlertLifeCycleObserver()
+	dispatcher := NewDispatcher(alerts, route, recorder, marker, timeout, nil, logger, m, observer)
+	go dispatcher.Run()
+	defer dispatcher.Stop()
+
+	// Create alerts. the dispatcher will automatically create the groups.
+	inputAlerts := []*types.Alert{
+		newAlert(model.LabelSet{"alertname": "OtherAlert", "cluster": "cc", "service": "dd"}),
+	}
+	err = alerts.Put(inputAlerts...)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Let alerts get processed.
+	for i := 0; len(recorder.Alerts()) != 1 && i < 10; i++ {
+		time.Sleep(200 * time.Millisecond)
+	}
+	require.Equal(t, 1, len(recorder.Alerts()))
+	require.Equal(t, inputAlerts[0].Fingerprint(), observer.AggrGroupAlerts[0].Fingerprint())
+	o, ok := notify.AlertLCObserver(dispatcher.ctx)
```

Comment:
can we create a fake observer for example increment a counter, then verify if the observer's function get called?

Reply:
Yes we already do that. In line 598 we create a fake observer and in like 616 we verify that the function was called by checking the recorded alert.